### DISCUSSION TITLE
Use multithreading to replace the current single background thread of horovod?

### DISCUSSION THREAD

**Discussion by Richie-yan** [horovod/horovod#2777] [Q&A]
Hi, does the team consider using multi-threading instead of the current single background thread of horovod?
Suppose that horovod is currently executing in the stage of waiting for the completion of the negotiated gradient calculation, but at this time, tensor_queue_ already has a gradient (and the gradient has been calculated), will this block the current execution of the gradient in tensor_queue_?
Does the team think about using multithreading? In other words, what more complicated operations might be introduced by using multithreading?

**Comment by romerojosh** [comment]
@Richie-yan For GPU execution at least, the collective operations are scheduled asynchronously by the background thread and a separate thread is spawned to trigger completion callbacks (look up `GPUOpContext::FinalizeGPUQueue` and its use in the code). This frees up the background thread to process new requests while the previous collective operations are pending or executing. 

With that said, the background thread can be held up in the `WAIT_FOR_DATA` phase on GPU, as the current implementation polls on CUDA events with the CPU: https://github.com/horovod/horovod/blob/2a775b2e059817093e476a5a44433426dda35ae5/horovod/common/operations.cc#L290-L309
This logic could be made more asynchronous in some paths by using `cudaStreamWaitEvent` on the Horovod CUDA stream, delegating all the waiting to the CUDA runtime and allowing the background thread to continue on.

**Comment by Richie-yan** [comment]
Hi, @romerojosh 
Thank you for your reply. Your suggestion is a good optimization idea for me. 
At the same time, I have done some other analysis here. Please give a review:

I split the stages of horovod and ran the bert-large model to get the following time proportions:
Negotiation: 31%, gradient waiting: 57%, host communication scheduling (host calls MemcpyInFusionBuffer, ncclAllReduce MemcpyOutFusionBuffer api): 12%

If the above three stages are optimized for multi-threading, the premise of gaining benefits is that the actual gradient communication time on the device side is lower than the total time consumption of the above three stages (that is, when the communication stream is idle, Optimize the above three stages to get benefits).

So I theoretically calculated that the above three stages took about 360.4ms in one train step, the actual gradient communication on the device side took about 305.4ms, and each step took 828.9ms, so the benefit was (360.4-305.4 ) / 828.9 = 0.066, that is, the benefits is about 6.6%.
Do you have any thoughts on this?
